crates/ty_python_semantic/src/types/diagnostic.rs

```diff
@@ -591,6 +592,27 @@ declare_lint! {
     }
 }
 
+declare_lint! {
+    /// ## What it does
+    /// Checks for the creation of invalid `TypeAliasType`s
+    ///
+    /// ## Why is this bad?
+    /// There are several requirements that you must follow when creating a `TypeAliasType`.
+    ///
+    /// ## Examples
+    /// ```python
+    /// from typing import TypeAliasType
+    ///
+    /// IntOrStr = TypeAliasType("IntOrStr", int | str)  # okay
+    /// NewAlias = TypeAliasType(get_name(), int)        # error: TypeAliasType name must be a string literal
```

Comment:
I wasn't sure if we should have the same strict requirements here as for legacy `TypeVar` constructions. I couldn't find anything in the spec, but pyright also emits diagnostics, if a `TypeAliasType` is not immediately assigned to a variable, for example.